src/main/java/com/amazon/aws/iot/greengrass/semver/Comparator.java

```diff
@@ -0,0 +1,177 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Translated from node-semver:
+ *
+ * The ISC License
+ * Copyright (c) Isaac Z. Schlueter and Contributors
+ *
+ * See com.amazon.aws.iot.greengrass.semver.THIRD-PARTY-LICENSE
+ */
+
+package com.amazon.aws.iot.greengrass.semver;
+
+import lombok.EqualsAndHashCode;
+
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@EqualsAndHashCode
+public class Comparator {
+  public final String value;
+  private String operator;
+  public SemVer semver;
+  public static final AnySemVer ANY = Comparator.AnySemVer.ANY;
+
+  public Comparator(String comp) {
+    this.parse(comp);
+
+    if (this.semver == ANY) {
+      this.value = "";
+    } else {
+      this.value = this.operator + this.semver.getVersion();
+    }
+  }
+
+  private void parse (String comp) {
+    Pattern r = SemVerRegexes.re.get(SemVerRegexes.t.get("COMPARATOR"));
+    Matcher m = r.matcher(comp);
+
+    if (!m.find()) {
+      throw new IllegalArgumentException("Invalid comparator " + comp);
+    }
+
+    this.operator = m.group(1) != null ? m.group(1) : "";
+    if (this.operator.equals("=")) {
+      this.operator = "";
+    }
+
+    // if it literally is just '>' or '' then allow anything.
+    if (m.group(2) == null) {
+      this.semver = ANY;
+    } else {
+      this.semver = new SemVer(m.group(2));
+    }
+  }
+
+  public boolean test (Object version) {
+    if (this.semver == ANY || version == ANY) {
+      return true;
+    }
+
+    if (version instanceof String) {
+      try {
+        version = new SemVer((String) version);
+      } catch (RuntimeException e) {
+        return false;
+      }
+    }
+
+    return cmp((SemVer) version, this.operator, this.semver);
+  }
+
+  public static boolean cmp(SemVer a, String op, SemVer b) {
+    switch (op) {
+      case "===":
+        return a.getVersion().equals(b.getVersion());
+      case "!==":
+        return !a.getVersion().equals(b.getVersion());
+      case "":
+      case "=":
+      case "==":
+        return eq(a, b);
+      case "!=":
+        return neq(a, b);
+      case ">":
+        return gt(a, b);
+      case ">=":
+        return gte(a, b);
+      case "<":
+        return lt(a, b);
+      case "<=":
+        return lte(a, b);
+      default:
+        throw new IllegalArgumentException("Invalid operator " + op);
+    }
+  }
+
+  public static boolean eq(SemVer a, SemVer b){
+    return a.compareTo(b) == 0;
+  }
+
+  public static boolean neq(SemVer a, SemVer b){
+    return a.compareTo(b) != 0;
+  }
+
+  public static boolean gt(SemVer a, SemVer b){
+    return a.compareTo(b) > 0;
+  }
+
+  public static boolean gte(SemVer a, SemVer b){
+    return a.compareTo(b) >= 0;
+  }
+
+  public static boolean lt(SemVer a, SemVer b){
+    return a.compareTo(b) < 0;
+  }
+
+  public static boolean lte(SemVer a, SemVer b){
+    return a.compareTo(b) <= 0;
+  }
+
+  public boolean intersects (Comparator comp) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 125-135 into a separate method.